src/bluesearch/database/mesh.py

```diff
@@ -0,0 +1,241 @@
+#  Blue Brain Search is a text mining toolbox focused on scientific use cases.
+#
+#  Copyright (C) 2022 Blue Brain Project, EPFL.
+#
+#  This program is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Utilities for handling MeSH topic data."""
+from __future__ import annotations
+
+import collections
+import json
+import logging
+import pathlib
+import re
+from collections.abc import Iterable
+from typing import Generator, TextIO
+
+logger = logging.getLogger(__name__)
+
+
+class MeSHTree:
+    """The hierarchical tree of MeSH topics.
+
+    The MeSH topic ontology forms a tree with most general topics at the
+    root and the most specific topics as the leafs. Here's a part of a MeSH
+    topic hierarchy
+
+    .. code-block:: text
+
+        Natural Science Disciplines [H01]
+            Biological Science Disciplines [H01.158]
+                Biology [H01.158.273]
+                    Botany [H01.158.273.118]
+                        Ethnobotany [H01.158.273.118.299]
+                        Pharmacognosy [H01.158.273.118.598]
+                            Herbal Medicine [H01.158.273.118.598.500]
+                    Cell Biology [H01.158.273.160]
+        ...
+
+    The full data can be found in the NLM's MeSH browser under
+    https://meshb.nlm.nih.gov/.
+
+    The topics are uniquely identified by their tree number (e.g. `H01.158`),
+    while the same topic label can appear in different places.
+
+    Parameters
+    ----------
+    tree_number_to_label
+        The MeSH tree data. This dictionary should have tree numbers
+        (e.g. `H01.158.273`) as keys, and topic labels (e.g. `Biology`)
+        as values.
+    """
+
+    def __init__(self, tree_number_to_label: dict[str, str]) -> None:
+        self.tree_number_to_label = tree_number_to_label
+        self.label_to_tree_numbers: dict[str, list[str]] = {}
+        for tree_number, label in tree_number_to_label.items():
+            if label not in self.label_to_tree_numbers:
+                self.label_to_tree_numbers[label] = []
```

Comment:
Maybe we could define `self.label_to_tree_numbers` as `defaultdict(list)`, what do you think ?

Reply:
Good idea! Added in 8e11d76.